### ISSUE TITLE
Vulkan Instance allocation

### ISSUE THREAD

**Issue by ShadowLordAlpha** [LWJGL/lwjgl3#187]
We should be able to allocate a `VkInstance` and pass it into the `vkCreateInstance()` method the same as is currently done with `VkInstanceCreateInfo` instead of it requiering a `PointerBuffer` as it currently does.

```
. . .
VkInstance vkInstance = VkInstance.malloc();
VK10.vkCreateInstance(vkInstanceCreateInfo, null, vkInstance);
. . .
```

**Comment by Spasi** [comment]
No, sorry, that would effectively make `VkInstance` objects mutable.

It would be possible to make `vkCreateInstance` return a `VkInstance`, but it already returns a value (the error code). An option would be to swap the `VkInstance` parameter and the return value like:

```java
IntBuffer pi = ...
VkInstance inst = vkCreateInstance(inst_info, null, pi);
int err = pi.get(0);
```

but I don't think that's any better.

**Comment by ShadowLordAlpha** [comment]
How would it make it more mutable that it already is? Currently you create a PointerBuffer outside of the object then just pass it in later so why not make it so that the object itself (or a static method) creates the pointer?

**Comment by ShadowLordAlpha** [comment]
So I should have added this but It would be more like a builder method than the current allocation method. so

```
VkInstance inst = VkInstance.malloc(inst_info);
```
where the method would automatically create an address for the object and then that address would be used in the `vkCreateInstance()` method instead of a `PointerBuffer` so almost nothing would change overall but it would make it look cleaner and less work for an end user. It would also be a bit more like the C code.

**Comment by Spasi** [comment]
> How would it make it more mutable that it already is?

Because between `VkInstance foo = new VkInstance()` and `vkCreateInstance(..., foo)`, you have an object that is invalid and cannot be used for anything. If `vkCreateInstance` fails, you still have that object. If `vkCreateInstance` succeeds, it would be possible to call it again and modify the existing object. It's bad design.

> So I should have added this but It would be more like a builder method than the current allocation method.

Best I can do is a factory method that calls `vkCreateInstance` for you. Something like (in the `VkInstance` class):

```java
public static VkInstance create(VkInstanceCreateInfo pCreateInfo, VkAllocationCallbacks pAllocator) {
	try ( MemoryStack stack = stackPush() ) {
		PointerBuffer pp = stack.mallocPointer(1);
		int err = vkCreateInstance(pCreateInfo, pAllocator, pp);
		if ( err != VK_SUCCESS )
			throw new RuntimeException("...");
		
		return new VkInstance(pp.get(0), pCreateInfo);
	}
}
```

and similarly for `VkDevice`. Error handling is still a problem though.

**Comment by ShadowLordAlpha** [comment]
> Because between VkInstance foo = new VkInstance() and vkCreateInstance(..., foo), you have an object that is invalid and cannot be used for anything. If vkCreateInstance fails, you still have that object. If vkCreateInstance succeeds, it would be possible to call it again and modify the existing object. It's bad design.

Thats how the current vulkan method works anyway and you can just keep using the same pointer over and over and over as well.

> Best I can do is a factory method that calls vkCreateInstance for you. Something like (in the VkInstance class)

It would probably be better to leave it up to the user but that is similar to what I currently do.

**Comment by httpdigest** [comment]
This is something which a layered library on top of LWJGL should be responsible for. Not LWJGL itself. IMO, LWJGL should be as lean and mean and close to native code as possible. Any fancy OO and convenience stuff like easy to use factory methods and ways to translate vk error codes to Java exceptions should be in another library.

**Comment by ShadowLordAlpha** [comment]
```Java
public static int vkCreateInstance(VkInstanceCreateInfo pCreateInfo, VkAllocationCallbacks pAllocator, VkInstance pInstance) {
		MemoryStack stack = stackGet(); int stackPointer = stack.getPointer();
		try {
		PointerBuffer pointer = stack.callocPointer(1);
		int code = nvkCreateInstance(pCreateInfo.address(), pAllocator == null ? NULL : pAllocator.address(), memAddress(pointer));
		if(code != VK_SUCCESS) {
			return code;
		} else {
			pInstance.setData(pointer.get(0), pCreateInfo);
		}
		} finally {
			stack.setPointer(stackPointer);
		}
	}
```
though `setData` could just always run but the way I did it is if there is an error code it just returns that code and `VkInstance` would be basically invalid anyway.

And the `VkInstance` code would be something like this.
```Java
. . .
/** Wraps a Vulkan instance handle. */
public class VkInstance extends DispatchableHandle {

	public VkInstance() {
		
	}

	/* 
	 * Allows the vkCreateInstance() method to set the proper data to the container
	 * without allow subclasses to access the method.
	 */
	void setData(long address, VkInstanceCreateInfo ci) {
		this.address = address;
		capabilities = getInstanceCapabilities(address, ci);
	}

	. . .
}
```
Though other container type objects used by the Vulkan bindings would also need to be modified as well. Also in the above samples I made it look a lot like what the code would look like if done currently though it might look better if the `VkInstance` was allocated and the handle placed there. I think it is just a pointer value so normally we would represent it as a long but the `VkCapabilities` are also stored in the Instance as well as in several other objects used by Vulkan.

now to go off my own topic but maybe something similar could be done with OpenAL to get rid of us needing to set the `ALCapabilities` when we change the context.

Also there are a few unneeded `null` checks in the generated code that uses `stack.UTF8()` as that method has a built in `null` check that it does first thing. (OpenAL code)

**Comment by Spasi** [comment]
> Also there are a few unneeded null checks in the generated code that uses stack.UTF8() as that method has a built in null check that it does first thing.

Thanks, fixed with d20f8f1a1bd0819be961ff2a2667bc823e8b8f6e.
